specs/experimental/gov-proposal-validator.md

```diff
@@ -0,0 +1,581 @@
+- `address[] optionsRecipients`: An address for each option to transfer funds to in case the option passes the voting.
+- `uint256[] optionsAmounts`: The amount to transfer for each option in case the option passes the voting.
+
+`approveProposal`
+
+Approves a proposal before being moved for voting, used by the top delegates.
+
+- MUST check if proposal hash corresponds to a valid proposal
+- MUST check if caller has enough voting power to call the function and approve a proposal
+  The voting power of the delegate is checked against the end of the last voting cycle from when
+  the proposal was submitted.
+- MUST check if caller has already approved the same proposal
+- MUST store the approval vote
+- MUST emit `ProposalApproved` when successfully called
+
+```solidity
+function approveProposal(bytes32 _proposalHash) external
+```
+
+Approving a funding proposal type requires extra attention to the budget amount and options, of the
+approval voting module, that were provided on the submission of the proposal. This should be handled
+by the Agora's UI.
+
+`moveToVote`
+
+Checks if the provided proposal is ready to move for voting. Based on the Proposal Type different checks are being
+validated. If all checks pass then `OptimismGovernor.proposeWithModule` is being called to forward the proposal for voting.
+
+- MUST create the `proposalHash` and check if it exists and is valid
+- Proposal MUST have gathered equal or more than the `requiredApprovals` defined for that type
+- MUST check if proposal has already moved for voting
+- MUST emit `ProposalMovedToVote` event
+
+For `ProtocolOrGovernorUpgrade` type:
+
+- MUST also check that is called by the proposer that submitted the proposal
+- `_optionsRecipients` and `_optionsAmounts` MUST be empty
+
+For `MaintenanceUpgradeProposals` type:
+
+- This type does not require any checks and is being forwarded to the Governor contracts, this should happen atomically.
+
+For `CouncilMemberElections`, `GovernanceFund` and `CouncilBudget` types:
+
+- Proposal MUST be moved to vote during a valid voting cycle
+- MUST check if the total amount of tokens that can possible be distributed during this voting cycle does not go over the
+  `VotingCycleData.votingCycleDistributionLimit`
+
+```solidity
+function moveToVote(
+    uint128 _criteriaValue,
+    address[] memory _optionsRecipients,
+    uint256[] memory _optionsAmounts,
+    string[] memory _optionDescriptions,
+    string memory _proposalDescription
+) external returns (uint256 proposalHash_)
+```
+
+`canApproveProposal`
+
+Returns true if a delegate address has enough voting power to approve a proposal.
+
+- Can be called by anyone
+- MUST return TRUE if the delegates' voting power is above the `minimumVotingPower`
+
+```solidity
+function canApproveProposal(address _delegate) public view returns (bool canSignOff_)
+```
+
+`setMinimumVotingPower`
+
+Sets the minimum voting power a delegate must have in order to be eligible to approve a proposal.
```

Comment:
Technically if the `canApproveProposal` check is based on the start of the voting cycle, wouldn't this need to be indexed by voting cycle or something?

If the associated complexity is significant, we can think about other options.  But if there's a voting cycle storage object, it might just be that you `setMinimumVotingPower(_power, _cycleId)` or something here instead